File: shared/lib/go/utils/circuit_breaker.go
package utils

import (
	"context"
	"errors"
	"sync"
	"time"
)

var (
	// ErrCircuitOpen is returned when the circuit breaker is open.
	ErrCircuitOpen = errors.New("circuit breaker is open")
	// ErrTooManyRequests is returned when too many requests are made in half-open state.
	ErrTooManyRequests = errors.New("too many requests")
)

// CircuitBreakerState represents the state of a circuit breaker.
type CircuitBreakerState int

const (
	// StateClosed means the circuit is closed (normal operation).
	StateClosed CircuitBreakerState = iota
	// StateOpen means the circuit is open (blocking requests).
	StateOpen
	// StateHalfOpen means the circuit is half-open (testing recovery).
	StateHalfOpen
)

// CircuitBreakerConfig configures a circuit breaker.
type CircuitBreakerConfig struct {
	MaxFailures     int           // Max failures before opening
	Timeout         time.Duration // Duration to stay open
	HalfOpenMaxReqs int           // Max requests in half-open state
	OnStateChange   func(from, to CircuitBreakerState)
}

// CircuitBreaker implements the circuit breaker pattern.
type CircuitBreaker struct {
	config        CircuitBreakerConfig
	state         CircuitBreakerState
	failures      int
	successes     int
	lastFailTime  time.Time
	halfOpenReqs  int
	mu            sync.RWMutex
}

// NewCircuitBreaker creates a new circuit breaker.
func NewCircuitBreaker(config CircuitBreakerConfig) *CircuitBreaker {
	if config.MaxFailures == 0 {
		config.MaxFailures = 5
	}
	if config.Timeout == 0 {
		config.Timeout = 30 * time.Second
	}
	if config.HalfOpenMaxReqs == 0 {
		config.HalfOpenMaxReqs = 3
	}

	return &CircuitBreaker{
		config: config,
		state:  StateClosed,
	}
}

// Execute executes the given function with circuit breaker protection.
func (cb *CircuitBreaker) Execute(ctx context.Context, fn func() error) error {
	if err := cb.beforeRequest(); err != nil {
		return err
	}

	err := fn()

	cb.afterRequest(err)
	return err
}

// GetState returns the current state of the circuit breaker.
func (cb *CircuitBreaker) GetState() CircuitBreakerState {
	cb.mu.RLock()
	defer cb.mu.RUnlock()
	return cb.state
}

// Reset resets the circuit breaker to closed state.
func (cb *CircuitBreaker) Reset() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	oldState := cb.state
	cb.state = StateClosed
	cb.failures = 0
	cb.successes = 0
	cb.halfOpenReqs = 0

	if cb.config.OnStateChange != nil && oldState != StateClosed {
		cb.config.OnStateChange(oldState, StateClosed)
	}
}

// beforeRequest checks if a request can be made.
func (cb *CircuitBreaker) beforeRequest() error {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case StateClosed:
		return nil

	case StateOpen:
		if time.Since(cb.lastFailTime) > cb.config.Timeout {
			cb.setState(StateHalfOpen)
			cb.halfOpenReqs = 1
			return nil
		}
		return ErrCircuitOpen

	case StateHalfOpen:
		if cb.halfOpenReqs >= cb.config.HalfOpenMaxReqs {
			return ErrTooManyRequests
		}
		cb.halfOpenReqs++
		return nil

	default:
		return nil
	}
}

// afterRequest records the result of a request.
func (cb *CircuitBreaker) afterRequest(err error) {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if err != nil {
		cb.onFailure()
	} else {
		cb.onSuccess()
	}
}

// onFailure handles a failed request.
func (cb *CircuitBreaker) onFailure() {
	cb.failures++
	cb.lastFailTime = time.Now()

	switch cb.state {
	case StateClosed:
		if cb.failures >= cb.config.MaxFailures {
			cb.setState(StateOpen)
		}

	case StateHalfOpen:
		cb.setState(StateOpen)
	}
}

// onSuccess handles a successful request.
func (cb *CircuitBreaker) onSuccess() {
	switch cb.state {
	case StateClosed:
		cb.failures = 0

	case StateHalfOpen:
		cb.successes++
		if cb.successes >= cb.config.HalfOpenMaxReqs {
			cb.setState(StateClosed)
			cb.failures = 0
			cb.successes = 0
		}
	}
}

// setState changes the circuit breaker state and triggers callback.
func (cb *CircuitBreaker) setState(newState CircuitBreakerState) {
	oldState := cb.state
	cb.state = newState
	cb.halfOpenReqs = 0

	if cb.config.OnStateChange != nil {
		cb.config.OnStateChange(oldState, newState)
	}
}

File: shared/lib/rust/common/src/circuit_breaker.rs
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::RwLock;

/// Circuit breaker state.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CircuitBreakerState {
    Closed,
    Open,
    HalfOpen,
}

/// Circuit breaker configuration.
#[derive(Debug, Clone)]
pub struct CircuitBreakerConfig {
    pub max_failures: usize,
    pub timeout: Duration,
    pub half_open_max_requests: usize,
}

impl Default for CircuitBreakerConfig {
    fn default() -> Self {
        CircuitBreakerConfig {
            max_failures: 5,
            timeout: Duration::from_secs(30),
            half_open_max_requests: 3,
        }
    }
}

/// Circuit breaker implementation.
pub struct CircuitBreaker {
    config: CircuitBreakerConfig,
    state: Arc<RwLock<CircuitBreakerInner>>,
}

struct CircuitBreakerInner {
    state: CircuitBreakerState,
    failures: usize,
    successes: usize,
    last_fail_time: Option<Instant>,
    half_open_requests: usize,
}

impl CircuitBreaker {
    /// Creates a new circuit breaker.
    pub fn new(config: CircuitBreakerConfig) -> Self {
        CircuitBreaker {
            config,
            state: Arc::new(RwLock::new(CircuitBreakerInner {
                state: CircuitBreakerState::Closed,
                failures: 0,
                successes: 0,
                last_fail_time: None,
                half_open_requests: 0,
            })),
        }
    }

    /// Executes a function with circuit breaker protection.
    pub async fn execute<F, T, E>(&self, f: F) -> Result<T, CircuitBreakerError<E>>
    where
        F: FnOnce() -> Result<T, E>,
    {
        self.before_request().await?;

        let result = f();

        self.after_request(result.is_ok()).await;

        result.map_err(CircuitBreakerError::Inner)
    }

    /// Executes an async function with circuit breaker protection.
    pub async fn execute_async<F, Fut, T, E>(&self, f: F) -> Result<T, CircuitBreakerError<E>>
    where
        F: FnOnce() -> Fut,
        Fut: std::future::Future<Output = Result<T, E>>,
    {
        self.before_request().await?;

        let result = f().await;

        self.after_request(result.is_ok()).await;

        result.map_err(CircuitBreakerError::Inner)
    }

    /// Returns the current state.
    pub async fn get_state(&self) -> CircuitBreakerState {
        self.state.read().await.state
    }

    /// Resets the circuit breaker to closed state.
    pub async fn reset(&self) {
        let mut inner = self.state.write().await;
        inner.state = CircuitBreakerState::Closed;
        inner.failures = 0;
        inner.successes = 0;
        inner.half_open_requests = 0;
    }

    async fn before_request(&self) -> Result<(), CircuitBreakerError<()>> {
        let mut inner = self.state.write().await;

        match inner.state {
            CircuitBreakerState::Closed => Ok(()),

            CircuitBreakerState::Open => {
                if let Some(last_fail) = inner.last_fail_time {
                    if last_fail.elapsed() > self.config.timeout {
                        inner.state = CircuitBreakerState::HalfOpen;
                        inner.half_open_requests = 1;
                        Ok(())
                    } else {
                        Err(CircuitBreakerError::Open)
                    }
                } else {
                    Err(CircuitBreakerError::Open)
                }
            }

            CircuitBreakerState::HalfOpen => {
                if inner.half_open_requests >= self.config.half_open_max_requests {
                    Err(CircuitBreakerError::TooManyRequests)
                } else {
                    inner.half_open_requests += 1;
                    Ok(())
                }
            }
        }
    }

    async fn after_request(&self, success: bool) {
        let mut inner = self.state.write().await;

        if success {
            self.on_success(&mut inner);
        } else {
            self.on_failure(&mut inner);
        }
    }

    fn on_success(&self, inner: &mut CircuitBreakerInner) {
        match inner.state {
            CircuitBreakerState::Closed => {
                inner.failures = 0;
            }

            CircuitBreakerState::HalfOpen => {
                inner.successes += 1;
                if inner.successes >= self.config.half_open_max_requests {
                    inner.state = CircuitBreakerState::Closed;
                    inner.failures = 0;
                    inner.successes = 0;
                }
            }

            _ => {}
        }
    }

    fn on_failure(&self, inner: &mut CircuitBreakerInner) {
        inner.failures += 1;
        inner.last_fail_time = Some(Instant::now());

        match inner.state {
            CircuitBreakerState::Closed => {
                if inner.failures >= self.config.max_failures {
                    inner.state = CircuitBreakerState::Open;
                }
            }

            CircuitBreakerState::HalfOpen => {
                inner.state = CircuitBreakerState::Open;
            }

            _ => {}
        }
    }
}

/// Circuit breaker error types.
#[derive(Debug, thiserror::Error)]
pub enum CircuitBreakerError<E> {
    #[error("Circuit breaker is open")]
    Open,

    #[error("Too many requests")]
    TooManyRequests,

    #[error("Inner error: {0}")]
    Inner(E),
}
